### PR TITLE
add support for lazy subscribers

### PR DESCRIPTION
**Pull request by mikeferguson**
This implements #111 for releases after Iron (where we have connect/disconnect callbacks, per https://github.com/ros2/rmw/issues/330)

### CODE REVIEW

**Comment by mikeferguson** [comment]
Also, not sure if rolling or ros2 is the right branch - they appear to be even with one another right now...

**Comment by clalancette** [comment]
> Also, not sure if rolling or ros2 is the right branch - they appear to be even with one another right now...

Yeah, it is a bit messy now.  `rolling` is the right branch, I retargeted to that.

**Comment by mikeferguson** [comment]
@clalancette so this can't go into Iron - but can it get merged into rolling now, or does it have to wait until some certain milestone between the Iron and J-turtle release?

**Comment by clalancette** [comment]
> @clalancette so this can't go into Iron - but can it get merged into rolling now, or does it have to wait until some certain milestone between the Iron and J-turtle release?

Rolling is completely open for business!

**Comment by ahcorde** [comment]
any updates here ?

**Comment by mikeferguson** [comment]
@ahcorde - this is basically waiting on a review from a maintainer (I've fully implemented the downstream stuff for image_proc and depth_image_proc, so I'm confident this is all we need to add)

**Review by ahcorde — APPROVED** [review]

**Comment by ahcorde** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/19151/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/13710/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/19886/)

**Pull request merged**
